include/drivers/i2c.h

```diff
@@ -212,6 +296,52 @@ static inline int z_impl_i2c_configure(struct device *dev, u32_t dev_config)
 	return api->configure(dev, dev_config);
 }
 
+/** @brief Perform asynchronous data transfer to another device.
+ *
+ * This routine provides a generic interface to perform asynchronous transfer of
+ * single message to another I2C device. Function can be called from interrupt
+ * context.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param msg Message to transfer.
+ * @param addr Address of the I2C target device.
+ * @param callback Callback called on transfer completion.
+ * @param user_data User data passed to the callback.
+ *
+ * @retval Non-negative on successful transfer scheduling.
+ * @retval -EBUSY if device is currently performing a transfer.
+ * @retval -EINVAL if any parameter is invalid.
+ */
+static inline int i2c_single_transfer(struct device *dev,
+					const struct i2c_msg *msg, u16_t addr,
+					i2c_transfer_callback_t callback,
```

Comment:
I would avoid this. Primarily because it's going to be called in interrupt context (I know, gpio does it, but we agreed about that when it was designed to do so). In SPI we use struct k_poll_signal instead for notifying when a call is done (the only difference being that the call does not return -EBUSY but waits until it gets its turn but that's another story). Using such kernel object is I think better since it provides more flexibility and robustness as well as avoiding to run a cb in interrupt context.

Reply:
If this is supposed to be an asynchronous API, I'd really like to see it use `async_notify` rather than assume a callback is the user's desired interface for results.

Reply:
async_notify would mean such async function would be usable only within the async library. It may not be what all users want. They might want to use a simpler system, thus why using k_poll_signal might be more relevant.

Reply:
`async_notify` (from #23229) is simply an abstraction that describes how the caller wants to be notified that an operation completes.  It allows callback, poll (signal)l, or spin-wait notifications, so can be used from a variety of contexts with or without `POLL` support.

Reply:
Sure, however result and flags attribute seem useless (signal owns already a result attribute, and for the cb it could just take a result as parameter). Would be worth having an intermediate structure just for signal or cb then.

Reply:
I think it makes sense to discuss the `async_notify` capability in #23229 rather than here.  There are reasons why those fields are the way they are.

Reply:
@tbursztyka 
> I would avoid this. Primarily because it's going to be called in interrupt context 

but that's the point. It is going to be called in the interrupt context and next transfer will also happen in the interrupt context. Operations are fully asynchronous. queue operation is scheduled from any context, then in transfer completion callback next one is taken from the sequence, if sequence is completed user is notified and queue is checked if there is another request pending and if yes next i2c transfer starts from interrupt context.

One thing to note is that `i2c_single_transfer` is not suppose to be used by the user, only by the `i2c_async` module. User can do that i.e. there is one sensor on the bus but then he gives up on access management thus scalability. Zephyr by default will use queue operation api (once/if ever all drivers switch to it).

@pabigot 
Regarding not using `async_notify` one of the reasons is the one mentioned above. It is not really intended to be used apart from i2c_async so it is tailored for that and adding async notify seems like overkill.

Reply:
I still think you don't need to run any cb on interrupt context, for i2c that is. If reaaaaally needed, then have a in-between solution with a union on a cb and a k_poll_signal

Reply:
a union like that would be `async_notify` then.

Reply:
async_notify holds 2 useless attributes (result and flags), it can be another one.

Reply:
They are not useless.  See https://github.com/zephyrproject-rtos/zephyr/pull/23229#discussion_r391546701.  Please provide feedback there if you are unconvinced so we don't have to track discussions in two places.

Reply:
flags are needed to distinguish between callback or signal. and i advocate for moving result into union (it's needed for spinwait mode): https://github.com/zephyrproject-rtos/zephyr/pull/23229#discussion_r390947643

Reply:
oh yes, missed that entirely. Well, do you really need a interrupt context cb then? :)

Reply:
yes, i do. I don't want dedicated thread to talk to i2c. If i have signal then i need thread context to wait for the completion and i want to be able to do that from any context (interrupts).